marklogic-data-hub/src/main/java/com/marklogic/hub/hubcentral/migration/HubCentralMigrator.java

```diff
@@ -0,0 +1,166 @@
+                }
+            }
+        }
+
+        if (atLeastOneEntityModelWasMigrated) {
+            logger.warn("Finished migrating entity models.");
+            logger.warn("Please examine your entity model files to verify that properties that were listed in the rangeIndex, pathRangeIndex, or elementRangeIndex arrays " +
+                    "now have \"facetable\":true in their property definition.\n");
+        } else {
+            logger.warn("No entity models required migration, so no project files were modified");
+        }
+    }
+
+    protected boolean entityModelRequiresMigration(ObjectNode entityModelNode) {
+        if (!entityModelValidForMigration(entityModelNode)) {
+            return false;
+        }
+        String firstLevelEntityTypeName = entityModelNode.get("info").get("title").asText();
+        entityModelNode = (ObjectNode) entityModelNode.get("definitions").get(firstLevelEntityTypeName);
+        return entityModelNode.get("rangeIndex") != null || entityModelNode.get("elementRangeIndex") != null ||
+                entityModelNode.get("pathRangeIndex") != null;
+    }
+
+    protected boolean entityModelValidForMigration(ObjectNode entityModelNode) {
+        if (!entityModelNode.has("info")) {
+            logger.warn("Info doesn't exist for the entity model and can not be migrated");
+            return false;
+        }
+
+        if (!entityModelNode.get("info").has("title")) {
+            logger.warn("Title doesn't exist for the entity model in the info and can not be migrated");
+            return false;
+        }
+
+        if (entityModelNode.get("info").get("title") == null || entityModelNode.get("info").get("title").asText().isEmpty()) {
+            logger.warn("Title is empty for the entity model in the info and can not be migrated");
+            return false;
+        }
+        String title = entityModelNode.get("info").get("title").asText();
+
+        if (!entityModelNode.has("definitions")) {
+            logger.warn("No definitions found in the entity model and can not be migrated");
+            return false;
+        }
+
+        if (entityModelNode.get("definitions").get(title) == null) {
+            logger.warn(format("entityType with title %s does not exist in the entityModel", title));
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean isStructuredTypeProperty(ObjectNode entityPropertyNode) {
+        // check for simple structured type property or simple relationship property
+        if (entityPropertyNode.get("datatype") == null && entityPropertyNode.get("$ref") != null) {
+            return true;
+        }
+
+        // check if structured type or relationship type property with array datatype
+        if (entityPropertyNode.get("datatype").asText().equals("array") && entityPropertyNode.get("items").get("$ref") != null) {
```

Comment:
We unfortunately need null checks on both "datatype" and "items" here - one may exist, but the other doesn't. We just can't trust that anything about a user's entity definition is valid (I've seen a few support tickets recently where entity models can't be loaded, and the cause is completely invalid ES syntax).